proto_parser/src/proto_parser.cc

```diff
@@ -0,0 +1,249 @@
+    std::make_shared<MessageValue>(name);
+  PopulateComments(last_field_loc, field_loc, message_val);
+  // Make sure the fields start from after the parent message.
+  last_field_loc = field_loc + 1;
+  PopulateFields(last_field_loc, tree, message, message_val);
+  return std::dynamic_pointer_cast<MessageValue>(message_val);
+}
+
+void ProtoParser::PopulateComments(int last_field_loc, 
+  int field_loc, std::shared_ptr<ProtoValue> message) {
+  std::string comments_above_field;
+  bool first_comment = false;
+  for(int i = last_field_loc; i < field_loc; ++i) {
+    // Make sure this line is not the closing bracket of a nested message
+    if(!first_comment) {
+      for (const char& c : lines_[i]) {
+        if (c != '}' && c != ' ' && c != '\t') {
+          first_comment = true;
+        }
+      }
+    }
+    if (first_comment) {
+      comments_above_field += lines_[i] + "\n";
+    }
+  }
+  // Figure out if the field is a string
+  std::string comment_behind_field;
+  std::string field_line = lines_[field_loc];
+  size_t begin_quote_idx = field_line.find('\"');
+  size_t pound_idx = field_line.find('#');
+  
+  if (begin_quote_idx != std::string::npos) {
+    size_t end_quote_idx = field_line.find("\"", begin_quote_idx + 1);
+    // Only if beginning and end quote exist is this a string.
+    if (end_quote_idx != std::string::npos) {
+      while(pound_idx != std::string::npos) {
+        if (pound_idx < begin_quote_idx) {
+          // If pound sign is before the string, then the string must be part of
+          // the comment eg. a: 2 #"hi"comment.
+          comment_behind_field = field_line.substr(pound_idx);
+          break;
+        }
+        else if (pound_idx > end_quote_idx) {
+          // If pound sign is after the string, then it must indicate beginning 
+          // of comment eg. a : "hi" #comment.
+          comment_behind_field = field_line.substr(pound_idx);
+          break;
+        }
+        pound_idx = field_line.find('#', pound_idx + 1);
+      }
+    }
+  }
+  else if (pound_idx != std::string::npos) {
+    comment_behind_field = field_line.substr(pound_idx);
+  }
+  message->SetCommentBehindField(comment_behind_field);
+  message->SetCommentAboveField(comments_above_field);
+}
+
+// TODO: Change return type to primitive value
+std::shared_ptr<PrimitiveValue> ProtoParser::CreatePrimitive(
+  const google::protobuf::Message& message, const FieldInfo& field, 
+  int last_field_loc) {
+  const google::protobuf::Reflection* reflection = message.GetReflection();
+  int field_loc = field.line_;
+  int index = field.index_;
+  const google::protobuf::FieldDescriptor* field_descriptor = 
+    field.field_descriptor_;
+  // String used for GetRepeatedStringReference.
+  std::string str;
+  std::shared_ptr<PrimitiveValue> primitive = std::make_shared<PrimitiveValue>(
+    field_descriptor->name());
+  PopulateComments(last_field_loc, field_loc, 
+    std::dynamic_pointer_cast<ProtoValue>(primitive));
```

Comment:
You don't need 
  std::dynamic_pointer_cast<ProtoValue>(primitive)

here.

Passing primitive can converts into ProtoValue automatically.